overlord/snapstate/component.go

```diff
@@ -0,0 +1,194 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapstate
+
+import (
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/overlord/snapstate/backend"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+// InstallComponentPath returns a set of tasks for installing a snap component
+// from a file path.
+//
+// Note that the state must be locked by the caller. The provided SideInfo can
+// contain just a name which results in local sideloading of the component, or
+// full metadata in which case the component will appear as installed from the
+// store.
+func InstallComponentPath(st *state.State, csi *snap.ComponentSideInfo, info *snap.Info,
+	path string, flags Flags) (*state.TaskSet, error) {
+	if info.RealName == "" {
+		return nil, fmt.Errorf(
+			"internal error: snap name to install component %q not provided",
+			path)
+	}
+
+	var snapst SnapState
+	err := Get(st, info.InstanceName(), &snapst)
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return nil, err
+	}
+
+	// Read ComponentInfo
+	compInfo, _, err := backend.OpenComponentFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check snap name matches
+	if compInfo.Component.SnapName != info.RealName {
+		return nil, fmt.Errorf(
+			"component snap name %q does not match real snap name %q",
+			compInfo.Component.SnapName, info.RealName)
+	}
+
+	snapsup := &SnapSetup{
+		Base:        info.Base,
+		SideInfo:    &info.SideInfo,
+		Channel:     info.Channel,
+		Flags:       flags.ForSnapSetup(),
+		Type:        info.Type(),
+		Version:     info.Version,
+		PlugsOnly:   len(info.Slots) == 0,
+		InstanceKey: info.InstanceKey,
+	}
+	compSetup := &ComponentSetup{
+		CompSideInfo: csi,
+		CompPath:     path,
+	}
+	// The file passed around is temporary, make sure it gets removed.
+	// TODO probably this should be part of a flags type in the future.
+	removeComponentPath := true
+	return doInstallComponent(st, &snapst, compSetup, snapsup, path, removeComponentPath, "")
+}
+
+// doInstallComponent assumes that the owner snap is already installed.
+func doInstallComponent(st *state.State, snapst *SnapState, compSetup *ComponentSetup,
+	snapsu *SnapSetup, path string, removeComponentPath bool, fromChange string) (*state.TaskSet, error) {
+
+	// TODO check for experimental flag that will hide temporarily components
+
+	snapSi := snapsu.SideInfo
+	compSi := compSetup.CompSideInfo
+
+	if snapst.IsInstalled() && !snapst.Active {
+		return nil, fmt.Errorf("cannot install component %q for disabled snap %q",
+			compSi.Component, snapSi.RealName)
+	}
+
+	// TODO extend conflict checks to components, this will check only for
+	// snaps conflicts (installation of a component as a snap gets the same
+	// conflicts as if we were installing the snap - which might be
+	// overkill and needs to be revisited).
```

Comment:
I think it's fine to consider an operation on a component a snap operation, especially once hooks get into play there is little reason to try to make the granularity smaller than that